architxt/database/__init__.py

```diff
@@ -0,0 +1,240 @@
+    """
+    table_info = {"Primary keys": {}, "Foreign keys": {}}
+
+    for table in tables:
+        primary_keys = inspector.get_pk_constraint(table)
+        table_info["Primary keys"][table] = primary_keys['constrained_columns']
+
+        foreign_keys = inspector.get_foreign_keys(table)
+        if foreign_keys:
+            table_info["Foreign keys"][table] = [
+                {'column': column, 'referred_table': fk['referred_table']}
+                for fk in foreign_keys
+                for column in fk['constrained_columns']
+            ]
+    return table_info
+
+
+def get_parent_tables(relations: dict[str, list[dict[str, str]]], tables: list[str]) -> set[str]:
+    """
+    Retrieve the parent tables in the database by identifying tables that are not referenced as foreign keys. Handle cycles by including all tables involved.
+
+    :param relations: A dictionary of foreign key relations.
+    :param tables: A list of table names.
+    :return: A set of parent table names.
+    """
+    childrens = {value["referred_table"] for info in relations.values() for value in info}
+
+    if not childrens:
+        return set(tables)
+
+    parents = set(relations.keys()) - childrens
+    cycle_tables = get_cycle_tables(relations)
+    if cycle_tables:
+        value = cycle_tables.pop()
+        parents.add(value)
+
+    return parents
+
+
+def get_cycle_tables(
+    relations: dict[str, list[dict[str, str]]],
+) -> set[str]:
+    """
+    Retrieve tables that are part of a cycle in the database relations.
+
+    :param relations: A dictionary of foreign key relations.
+    :return: A set of table names that are part of a cycle.
+    """
+    cycle_tables = set()
+    for table, info in relations.items():
+        for value in info:
+            if value['referred_table'] == table:
+                continue
+            if value['referred_table'] in relations and table in [
+                x['referred_table'] for x in relations[value['referred_table']]
+            ]:
+                cycle_tables.add(table)
+                cycle_tables.add(value['referred_table'])
+    return cycle_tables
+
+
+def process_table_relations(
+    table: str,
+    relations: dict[str, list[dict[str, str]]],
+    session: sessionmaker,
+    inspector: inspect,
+    primary_keys: dict[str, list[str]],
+) -> list[Tree]:
+    """
+    Process the relations of a given table, retrieve data, and construct tree representations.
+
+    :param table: Name of the table to process.
+    :param relations: Dictionary of foreign key relations.
+    :param session: SQLAlchemy session object.
+    :param inspector: SQLAlchemy inspector object.
+    :param primary_keys: Dictionary of primary keys for tables.
+    :return: A list of trees representing the relations and data for the table.
+    """
+    trees = []
+
+    # Retrieve data from the table
+    info = session.execute(text(f"SELECT * FROM '{table}'")).fetchmany(3)
```

Comment:
By using SQLAlchemy reflection, you can leverage the programmatic query interface: https://docs.sqlalchemy.org/en/20/orm/session_basics.html#querying.

While this query is simple, using the query interface generally ensures compatibility across different SQL dialects.

Additionally, avoid string concatenation or formatting in SQL queries to prevent SQL injection. If you need to define a query manually, always use parameterized queries. They also improve performance by allowing execution plan caching.